File: dfs/paths/binary_tree_paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
    @paths = []
    dfs(root)
    @paths
end

def dfs(node, path = [])
    return if !node  # Corner case when root itself is null, no leaf node
    path = path + [node.val]

    # As soon as we encounter a leaf node, push the path so far into paths list
    if !node.left && !node.right
        @paths << path.join('->')
    # For all other cases, just traverse down left and right children
    else
        dfs(node.left, path)
        dfs(node.right, path)
    end
end

# 257. Binary Tree Paths
# https://leetcode.com/problems/binary-tree-paths/description/

# Time: O(n)
# Space: Avg O(log(n)), Worst case O(n), Based on the depth of the tree for call stack.

File: dfs/paths/binary_tree_paths.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        self.paths = []
        self.dfs(root, [])
        return self.paths

    def dfs(self, root, path):
        if not root: return
        path = path + [str(root.val)]

        if self.is_leaf(root):  self.paths.append("->".join(path))
        else:                   self.dfs(root.left, path); self.dfs(root.right, path)

    def is_leaf(self, node):
        return not node.left and node.right



# 257. Binary Tree Paths
# https://leetcode.com/problems/binary-tree-paths/description/

# Intuition:
# - Again here only paths from root to leaf are considered and not root to None nodes
#   So we need to explicitly check for leaf nodes as the terminating conditions
# - Same like min depth of binary tree and 112. Path Sum

# Time: O(n)
# Space: Avg O(log(n)), Worst case O(n), Based on the depth of the tree for call stack.
